src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -24,120 +28,144 @@ public static int CacheSize
                 if (value < 0)
                     throw new ArgumentOutOfRangeException(nameof(value));
 
-                s_cacheSize = value;
-                if (s_livecode.Count > s_cacheSize)
+                lock (s_livecode)
                 {
-                    lock (s_livecode)
+                    s_cacheSize = value;  // not to allow other thread to change it while we use cache
+                    while (s_livecode.Count > s_cacheSize)
                     {
-                        while (s_livecode.Count > s_cacheSize)
-                            s_livecode.RemoveLast();
+                        CachedCodeEntry last = s_livecode_last;
```

Comment:
Not suggesting you change it now, but if you walked from the start of the list downwards, you could trim off the excess by just adjusting nodes either side of the boundary, without touching every item after that point?

Of course depending on what the new value is walking from the start could be slower than walking from the end.

Reply:
Eh, you still need to at least read them all to remove from the dictionary.

Reply:
Yes, needs to clear the dictionary still. Though we can avoid some operations for the linked list. I guess not so important.